### DISCUSSION TITLE
How do i get the folder name from toxinidir (full path) variable?

### DISCUSSION THREAD

**Discussion by abhinav-badola** [tox-dev/tox#2963] [Q&A] [answered]
I am new to tox and currently implementing tox in a pipeline. We use pytest for testing the code and want to store the reports of the unit tests in a different folder than the source code. I have been trying to play around with {toxinidir} variable but it returns the full path of the directory and for the report path, i want to calculate the relative path from the folder in which tox.ini file is placed. 

I also tried to do tox -x or --override to override the set variables in the tox.ini file but doesn't look like i am using it correctly.

I have attached a sample of the tox.ini file i am using.
```
[tox]
    skipsdist = True
    passenv = *
    envlist = py38,py39
[testenv]
setenv =
    AWS_DEFAULT_REGION = eu-west-1
    PYTHONPATH = $PYTHONPATH{:}{toxinidir}{:}../../../source/layers/xyz
    REPORT_PREFIX = 'DUMMY'
deps=
    pytest
    boto3
    moto
    mock
commands =
    pytest {posargs} -v --cov=tests --junitxml=../../../reports/{env:REPORT_PREFIX}-reports.xml
```

Any help/pointers would be appreciated. :)

TIA,
Abhinav

**Comment by jugmac00** [comment]
What is the issue you experience when combining `toxinidir` and a relative path?

**Comment by abhinav-badola** [comment]
For instance my toxinidir is:  /codebuild/output/tmp/.builds/region/ABC_001
but i want to store the report at path: /codebuild/output/tmp/reports/
Report name should be: ABC_001-reports.xml

When i use toxinidir as per the example tox.ini file:
    pytest {posargs} -v --cov=tests --junitxml=../../../reports{toxinidir}-reports.xml
would not resolve in the report path but i am ending up with:
pytest {posargs} -v --cov=tests --junitxml=/codebuild/output/tmp/reports/codebuild/output/tmp/.builds/region/ABC_001-reports.xml

So i want to extract the folder name ABC_001 from the path of toxinidir variable and then use it in the pytest command

**Comment by jugmac00** [comment]
You need to start with `toxinidir` when assembling a path.

e.g. and untested `pytest {posargs} -v --cov=tests --junitxml={toxinidir}../../../{env:REPORT_PREFIX}-reports.xml`

You can easily try things out in your terminal... e.g. for me `cd /home/jugmac00/..` results in `/home`.

**Comment by abhinav-badola** [comment]
Yes, that works but i still want to use the foldername also as the report-prefix:

Already known facts:
toxinidir = /codebuild/output/tmp/.builds/region/ABC_001
reportdir =  /codebuild/output/tmp/reports/

and I want to extract the directory name (ABC_001) from toxinidir full path and then use it as
reportfilename = ABC_001-reports.xml
reportfullpath =  /codebuild/output/tmp/reports/ABC_001-reports.xml

I am mainly  struggling with how to calculate the reportfilename from toxinidir variable.

**Comment by jugmac00** [comment, marked as answer]
I think you could also wrap your command in a bashfile and there you can use `basename` to get the directory name.

```
❯ basename /home/jugmac00
jugmac00
```

**Reply by abhinav-badola** [reply]
Yes, that works to fetch the filename but i was not able to make it work if i use it as a script within tox.in in one line.

May be because my understanding of tox.ini is limited but
setting a variable in command section doesn't work for me and also tried something like:

pytest {posargs} -v --cov=tests --junitxml=$(basename {toxinidir})-reports.xml

which works on shell but doesn't work when i execute via tox
as

tox -c tox.ini

**Comment by jugmac00** [comment]
I was referring to create a real bash file, and then executing it in the commands section.

ie create a test.sh file, put all your manipulations in there, and then call it in the commands section.

```
commands = test.sh
```

**Comment by abhinav-badola** [comment]
Thanks a lot @jugmac00 for the the help, it works indeed. But i was thinking of a more elegant solution to handle all this inside the tox.ini file itself. I have placed test.sh in a common directory rather than copying it to all the build directories for each python package we generate. 

Only downside of this solution is there is no one place where all config was stored as in (tox.ini) for one particular python package/source code and this is just a simple version but we will have to customise (test.sh) further for some edge cases. Would have been nice if all this could have been done in tox.ini.



test.sh
```
pytest -v --cov=tests --junitxml=../../../reports/$(basename $1)-$2-reports.xml
```
tox.ini
```
[tox]
skipsdist = True
passenv = *
envlist = py38,py39
[testenv]
setenv =
    AWS_DEFAULT_REGION = eu-west-1
    PYTHONPATH = $PYTHONPATH{:}{toxinidir}{:}../../../source/layers/xyz
deps=
    pytest
    boto3
    moto
    pytest-cov
    mock
    inflection
allowlist_externals=
    bash
commands =
    bash ../../../build/test.sh {toxinidir} {envname}
```


Also, as i am invoking tox via my python script for each python package like:
I am using tox 4.4.7

```
import tox

args = ['-c', f'{self.build_path}/{TOX_INI_FILE}', '-v']
results = tox.main(args)

```
is it also possible to somehow pass that information  when calling tox.main() which can be consumed by the pytest command in tox.ini file?
